lib/digest/cryb_sha1.3
.\"-
.\" Copyright (c) 2015 The University of Oslo
.\" Copyright (c) 2016-2022 Dag-Erling Smørgrav
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote
.\"    products derived from this software without specific prior written
.\"    permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd May 15, 2022
.Dt CRYB_SHA1 3
.Os
.Sh NAME
.Nm cryb_sha1_init ,
.Nm cryb_sha1_update ,
.Nm cryb_sha1_final ,
.Nm cryb_sha1_complete
.Nd Secure Hash Algorithm 1
.Sh LIBRARY
.Lb libcryb-digest
.Sh SYNOPSIS
.In stddef.h
.In stdint.h
.In cryb/sha1.h
.Ft void
.Fn cryb_sha1_init "cryb_sha1_ctx *context"
.Ft void
.Fn cryb_sha1_update "cryb_sha1_ctx *context" "const void *data" "size_t len"
.Ft void
.Fn cryb_sha1_final "cryb_sha1_ctx *context" "uint8_t *digest"
.Ft void
.Fn cryb_sha1_complete "const void *data" "size_t len" "uint8_t *digest"
.Sh DESCRIPTION
The
.Nm cryb_sha1
family of functions implements the NIST SHA-1 message digest algorithm
as described in FIPS 180-4.
.Pp
The
.Fn cryb_sha1_init
function initializes the context structure pointed to by
.Va context .
It is the caller's responsibility to allocate this structure.
.Pp
The
.Fn cryb_sha1_update
function hashes the next
.Va len
bytes of data pointed to by the
.Va data
pointer into the given hash context.
.Pp
The
.Fn cryb_sha1_final
function finalizes the computation and writes the resulting message
digest to the caller-provided buffer pointed to by
.Va digest ,
which must be at least
.Dv SHA1_DIGEST_LEN
bytes long.
It is the caller's responsibility to dispose of the context structure
after calling
.Fn cryb_sha1_final .
.Pp
The
.Fn cryb_sha1_complete
function is a shortcut to calling
.Fn cryb_sha1_init ,
.Fn cryb_sha1_update ,
and
.Fn cryb_sha1_final
when the entire message is available up front in a single contiguous
buffer.
.Sh IMPLEMENTATION NOTES
The
.In cryb/sha1.h
header provides macros which allows these functions and the context
type to be referred to without their
.Dq Li cryb_
prefix.
.Pp
The SHA-1 message digest algorithm has been proven to be vulnerable to
collision attacks and should not be used for cryptographic purposes.
.Sh SEE ALSO
.Xr cryb_digest 3 ,
.Xr cryb_sha224 3 ,
.Xr cryb_sha256 3 ,
.Xr cryb_sha384 3 ,
.Xr cryb_sha512 3
.Sh REFERENCES
.Rs
.%Q National Institute of Standards and Technology
.%R Secure Hash Standard (SHS) (FIPS PUB 180-4)
.%D March 2012
.Re
.Sh AUTHORS
.An -nosplit
These functions and this manual page were written for the University
of Oslo by
.An Dag-Erling Sm\(/orgrav Aq Mt d.e.smorgrav@usit.uio.no .


debian/changelog
cryb-to (0.20190326-1) unstable; urgency=low

  * New alpha release

 -- Dag-Erling Smørgrav <des@des.no>  Tue, 26 Mar 2019 12:24:03 +0100

cryb-to (0.20181115-1) unstable; urgency=low

  * New alpha release

 -- Dag-Erling Smørgrav <des@des.no>  Thu, 15 Nov 2018 18:58:09 +0100

cryb-to (0.20180513-1) unstable; urgency=low

  * New alpha release

 -- Dag-Erling Smørgrav <des@des.no>  Sun, 13 May 2018 16:40:11 +0200

cryb-to (0.20180426-1) unstable; urgency=low

  * New alpha release

 -- Dag-Erling Smørgrav <des@des.no>  Thu, 26 Apr 2018 22:04:25 +0200

cryb-to (0.20170904-1) unstable; urgency=low

  * New alpha release

 -- Dag-Erling Smørgrav <des@des.no>  Mon, 04 Sep 2017 21:45:06 +0200

cryb-to (0.20170502-1) unstable; urgency=low

  * New alpha release

 -- Dag-Erling Smørgrav <des@des.no>  Tue, 02 May 2017 13:16:49 +0200

cryb-to (0.20170406-1) unstable; urgency=low

  * New alpha release

 -- Dag-Erling Smørgrav <des@des.no>  Wed, 06 Apr 2017 21:29:46 +0200

cryb-to (0.20170322-1) unstable; urgency=low

  * New alpha release

 -- Dag-Erling Smørgrav <des@des.no>  Wed, 22 Mar 2017 22:00:35 +0100

cryb-to (0.20170307-1) unstable; urgency=low

  * Initial release

 -- Dag-Erling Smørgrav <des@des.no>  Wed, 09 Mar 2017 22:30:06 +0100


lib/digest/cryb_sha1.c
/*-
 * Copyright (c) 2012 The University of Oslo
 * Copyright (c) 2012-2016 Dag-Erling Smørgrav
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior written
 *    permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "cryb/impl.h"

#include <stdint.h>
#include <string.h>

#include <cryb/bitwise.h>
#include <cryb/endian.h>
#include <cryb/memset_s.h>

#include <cryb/sha1.h>

static uint32_t sha1_h[5] = {
	0x67452301U, 0xefcdab89U, 0x98badcfeU, 0x10325476U, 0xc3d2e1f0U,
};

static uint32_t sha1_k[4] = {
	0x5a827999U, 0x6ed9eba1U, 0x8f1bbcdcU, 0xca62c1d6U,
};

void
sha1_init(sha1_ctx *ctx)
{

	memset(ctx, 0, sizeof *ctx);
	memcpy(ctx->h, sha1_h, sizeof ctx->h);
}

#define sha1_ch(x, y, z)	((x & y) ^ (~x & z))
#define sha1_parity(x, y, z)	((x ^ y ^ z))
#define sha1_maj(x, y, z)	(((x & y) ^ (x & z) ^ (y & z)))
#define sha1_step(t, a, f, e, w)					\
	do {								\
		uint32_t T = rol32(a, 5) + f + e + sha1_k[t/20] + w[t];	\
		e = d;							\
		d = c;							\
		c = rol32(b, 30);					\
		b = a;							\
		a = T;							\
	} while (0)

static void
sha1_compute(sha1_ctx *ctx, const uint8_t *block)
{
	uint32_t w[80], a, b, c, d, e;
	unsigned int i;

	be32decv(w, block, 16);
	for (i = 16; i < 80; ++i) {
		w[i] = w[i-3] ^ w[i-8] ^ w[i-14] ^ w[i-16];
		w[i] = rol32(w[i], 1);
	}
	a = ctx->h[0];
	b = ctx->h[1];
	c = ctx->h[2];
	d = ctx->h[3];
	e = ctx->h[4];

	sha1_step( 0, a, sha1_ch(b, c, d), e, w);
	sha1_step( 1, a, sha1_ch(b, c, d), e, w);
	sha1_step( 2, a, sha1_ch(b, c, d), e, w);
	sha1_step( 3, a, sha1_ch(b, c, d), e, w);
	sha1_step( 4, a, sha1_ch(b, c, d), e, w);
	sha1_step( 5, a, sha1_ch(b, c, d), e, w);
	sha1_step( 6, a, sha1_ch(b, c, d), e, w);
	sha1_step( 7, a, sha1_ch(b, c, d), e, w);
	sha1_step( 8, a, sha1_ch(b, c, d), e, w);
	sha1_step( 9, a, sha1_ch(b, c, d), e, w);
	sha1_step(10, a, sha1_ch(b, c, d), e, w);
	sha1_step(11, a, sha1_ch(b, c, d), e, w);
	sha1_step(12, a, sha1_ch(b, c, d), e, w);
	sha1_step(13, a, sha1_ch(b, c, d), e, w);
	sha1_step(14, a, sha1_ch(b, c, d), e, w);
	sha1_step(15, a, sha1_ch(b, c, d), e, w);
	sha1_step(16, a, sha1_ch(b, c, d), e, w);
	sha1_step(17, a, sha1_ch(b, c, d), e, w);
	sha1_step(18, a, sha1_ch(b, c, d), e, w);
	sha1_step(19, a, sha1_ch(b, c, d), e, w);

	sha1_step(20, a, sha1_parity(b, c, d), e, w);
	sha1_step(21, a, sha1_parity(b, c, d), e, w);
	sha1_step(22, a, sha1_parity(b, c, d), e, w);
	sha1_step(23, a, sha1_parity(b, c, d), e, w);
	sha1_step(24, a, sha1_parity(b, c, d), e, w);
	sha1_step(25, a, sha1_parity(b, c, d), e, w);
	sha1_step(26, a, sha1_parity(b, c, d), e, w);
	sha1_step(27, a, sha1_parity(b, c, d), e, w);
	sha1_step(28, a, sha1_parity(b, c, d), e, w);
	sha1_step(29, a, sha1_parity(b, c, d), e, w);
	sha1_step(30, a, sha1_parity(b, c, d), e, w);
	sha1_step(31, a, sha1_parity(b, c, d), e, w);
	sha1_step(32, a, sha1_parity(b, c, d), e, w);
	sha1_step(33, a, sha1_parity(b, c, d), e, w);
	sha1_step(34, a, sha1_parity(b, c, d), e, w);
	sha1_step(35, a, sha1_parity(b, c, d), e, w);
	sha1_step(36, a, sha1_parity(b, c, d), e, w);
	sha1_step(37, a, sha1_parity(b, c, d), e, w);
	sha1_step(38, a, sha1_parity(b, c, d), e, w);
	sha1_step(39, a, sha1_parity(b, c, d), e, w);

	sha1_step(40, a, sha1_maj(b, c, d), e, w);
	sha1_step(41, a, sha1_maj(b, c, d), e, w);
	sha1_step(42, a, sha1_maj(b, c, d), e, w);
	sha1_step(43, a, sha1_maj(b, c, d), e, w);
	sha1_step(44, a, sha1_maj(b, c, d), e, w);
	sha1_step(45, a, sha1_maj(b, c, d), e, w);
	sha1_step(46, a, sha1_maj(b, c, d), e, w);
	sha1_step(47, a, sha1_maj(b, c, d), e, w);
	sha1_step(48, a, sha1_maj(b, c, d), e, w);
	sha1_step(49, a, sha1_maj(b, c, d), e, w);
	sha1_step(50, a, sha1_maj(b, c, d), e, w);
	sha1_step(51, a, sha1_maj(b, c, d), e, w);
	sha1_step(52, a, sha1_maj(b, c, d), e, w);
	sha1_step(53, a, sha1_maj(b, c, d), e, w);
	sha1_step(54, a, sha1_maj(b, c, d), e, w);
	sha1_step(55, a, sha1_maj(b, c, d), e, w);
	sha1_step(56, a, sha1_maj(b, c, d), e, w);
	sha1_step(57, a, sha1_maj(b, c, d), e, w);
	sha1_step(58, a, sha1_maj(b, c, d), e, w);
	sha1_step(59, a, sha1_maj(b, c, d), e, w);

	sha1_step(60, a, sha1_parity(b, c, d), e, w);
	sha1_step(61, a, sha1_parity(b, c, d), e, w);
	sha1_step(62, a, sha1_parity(b, c, d), e, w);
	sha1_step(63, a, sha1_parity(b, c, d), e, w);
	sha1_step(64, a, sha1_parity(b, c, d), e, w);
	sha1_step(65, a, sha1_parity(b, c, d), e, w);
	sha1_step(66, a, sha1_parity(b, c, d), e, w);
	sha1_step(67, a, sha1_parity(b, c, d), e, w);
	sha1_step(68, a, sha1_parity(b, c, d), e, w);
	sha1_step(69, a, sha1_parity(b, c, d), e, w);
	sha1_step(70, a, sha1_parity(b, c, d), e, w);
	sha1_step(71, a, sha1_parity(b, c, d), e, w);
	sha1_step(72, a, sha1_parity(b, c, d), e, w);
	sha1_step(73, a, sha1_parity(b, c, d), e, w);
	sha1_step(74, a, sha1_parity(b, c, d), e, w);
	sha1_step(75, a, sha1_parity(b, c, d), e, w);
	sha1_step(76, a, sha1_parity(b, c, d), e, w);
	sha1_step(77, a, sha1_parity(b, c, d), e, w);
	sha1_step(78, a, sha1_parity(b, c, d), e, w);
	sha1_step(79, a, sha1_parity(b, c, d), e, w);

	ctx->h[0] += a;
	ctx->h[1] += b;
	ctx->h[2] += c;
	ctx->h[3] += d;
	ctx->h[4] += e;
}

void
sha1_update(sha1_ctx *ctx, const void *buf, size_t len)
{
	size_t copylen;

	while (len) {
		if (ctx->blocklen > 0 || len < sizeof ctx->block) {
			copylen = sizeof ctx->block - ctx->blocklen;
			if (copylen > len)
				copylen = len;
			memcpy(ctx->block + ctx->blocklen, buf, copylen);
			ctx->blocklen += copylen;
			if (ctx->blocklen == sizeof ctx->block) {
				sha1_compute(ctx, ctx->block);
				ctx->blocklen = 0;
			}
		} else {
			copylen = sizeof ctx->block;
			sha1_compute(ctx, buf);
		}
		ctx->bitlen += copylen * 8;
		buf += copylen;
		len -= copylen;
	}
}

void
sha1_final(sha1_ctx *ctx, uint8_t *digest)
{

	ctx->block[ctx->blocklen++] = 0x80;
	memset(ctx->block + ctx->blocklen, 0,
	    sizeof ctx->block - ctx->blocklen);
	if (ctx->blocklen > 56) {
		sha1_compute(ctx, ctx->block);
		ctx->blocklen = 0;
		memset(ctx->block, 0, sizeof ctx->block);
	}
	be64enc(ctx->block + 56, ctx->bitlen);
	sha1_compute(ctx, ctx->block);
	be32encv(digest, ctx->h, 5);
	memset_s(ctx, 0, sizeof *ctx, sizeof *ctx);
}

void
sha1_complete(const void *buf, size_t len, uint8_t *digest)
{
	sha1_ctx ctx;

	sha1_init(&ctx);
	sha1_update(&ctx, buf, len);
	sha1_final(&ctx, digest);
}

digest_algorithm sha1_digest = {
	.name			 = "sha1",
	.contextlen		 = sizeof(sha1_ctx),
	.blocklen		 = SHA1_BLOCK_LEN,
	.digestlen		 = SHA1_DIGEST_LEN,
	.init			 = (digest_init_func)sha1_init,
	.update			 = (digest_update_func)sha1_update,
	.final			 = (digest_final_func)sha1_final,
	.complete		 = (digest_complete_func)sha1_complete,
};
